src/main/java/picard/illumina/NewIlluminaBasecallsConverter.java

```diff
@@ -0,0 +1,393 @@
+     * @param numProcessors            Controls number of threads.  If <= 0, the number of threads allocated is
+     *                                 available cores - numProcessors.
+     * @param firstTile                (For debugging) If non-null, start processing at this tile.
+     * @param tileLimit                (For debugging) If non-null, process no more than this many tiles.
+     * @param outputRecordComparator   For sorting output records within a single tile.
+     * @param codecPrototype           For spilling output records to disk.
+     * @param outputRecordClass        Inconveniently needed to create SortingCollections.
+     * @param includeNonPfReads        If true, will include ALL reads (including those which do not have PF set)
+     * @param ignoreUnexpectedBarcodes If true, will ignore reads whose called barcode is not found in barcodeRecordWriterMap,
+     * @param maxNoCalls               The maximum number of no calls to allow in a match.
+     * @param maxMismatches            The maximum number of mismatched calls to allow in a match.
+     * @param minMismatchDelta         The minimum mismatch difference between the best and second best matches.
+     * @param minimumBaseQuality       The minimum base quality to allow for a matching base call.
+     * @param metrics                  The metrics output for barcode metrics.
+     * @param metricsFile              The metrics file to write out the barcode metrics to.
+     */
+    public NewIlluminaBasecallsConverter(final File basecallsDir, File barcodesDir, final int lane,
+                                         final ReadStructure readStructure,
+                                         final Map<String, ? extends ConvertedClusterDataWriter<CLUSTER_OUTPUT_RECORD>> barcodeRecordWriterMap,
+                                         final boolean demultiplex,
+                                         final int maxReadsInRamPerTile,
+                                         final List<File> tmpDirs, final int numProcessors,
+                                         final Integer firstTile,
+                                         final Integer tileLimit,
+                                         final Comparator<CLUSTER_OUTPUT_RECORD> outputRecordComparator,
+                                         final SortingCollection.Codec<CLUSTER_OUTPUT_RECORD> codecPrototype,
+                                         final Class<CLUSTER_OUTPUT_RECORD> outputRecordClass,
+                                         final BclQualityEvaluationStrategy bclQualityEvaluationStrategy,
+                                         final boolean applyEamssFiltering, final boolean includeNonPfReads,
+                                         final boolean ignoreUnexpectedBarcodes, int maxNoCalls, int maxMismatches,
+                                         int minMismatchDelta, int minimumBaseQuality, MetricsFile<BarcodeMetric, Integer> metrics, File metricsFile) {
```

Comment:
Is there a reason why we use `final` for parameters and locals? I think `final` is great for fields (and its use on methods and classes has a different semantic meaning) but I'm not convinced it buys us much on parameters and locals. If methods are a reasonable size it shouldn't be hard to tell if the code is assigning to a parameter or field.